// react-flagkit/index.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('react'), require('prop-types')) :
	typeof define === 'function' && define.amd ? define(['exports', 'react', 'prop-types'], factory) :
	(factory((global['react-flagkit'] = {}),global.React,global.PropTypes));
}(this, (function (exports,React,propTypes) { 'use strict';

React = React && React.hasOwnProperty('default') ? React['default'] : React;
propTypes = propTypes && propTypes.hasOwnProperty('default') ? propTypes['default'] : propTypes;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Displays an SVG `img` tag with a country flag.
 * @public
 * @param       {string} country
 * @param       {string} role
 * @param       {number} size
 * @returns     {object}
 * @constructor
 */
function Flag(props) {
  var country = props.country,
      role = props.role,
      size = props.size,
      otherProps = _objectWithoutProperties(props, ['country', 'role', 'size']);

  var flagSrc = require('./img/SVG/' + country + '.svg');
  return React.createElement('img', _extends({
    src: flagSrc,
    role: role,
    alt: country + ' Flag',
    height: size,
    width: size
  }, otherProps));
}

Flag.defaultProps = {
  role: 'img',
  size: 24
};

exports['default'] = Flag;

Object.defineProperty(exports, '__esModule', { value: true });

})));
